builtin/logical/pki/cert_util.go

```diff
@@ -255,16 +608,261 @@ func createCertificate(creationInfo *certCreationBundle) (*certutil.ParsedCertBu
 	if creationInfo.Usage&codeSigningUsage != 0 {
 		certTemplate.ExtKeyUsage = append(certTemplate.ExtKeyUsage, x509.ExtKeyUsageCodeSigning)
 	}
+	if creationInfo.Usage&emailProtectionUsage != 0 {
+		certTemplate.ExtKeyUsage = append(certTemplate.ExtKeyUsage, x509.ExtKeyUsageEmailProtection)
+	}
+
+	certTemplate.IssuingCertificateURL = creationInfo.URLs.IssuingCertificates
+	certTemplate.CRLDistributionPoints = creationInfo.URLs.CRLDistributionPoints
+	certTemplate.OCSPServer = creationInfo.URLs.OCSPServers
+
+	var certBytes []byte
+	if creationInfo.SigningBundle != nil {
+		switch creationInfo.SigningBundle.PrivateKeyType {
+		case certutil.RSAPrivateKey:
+			certTemplate.SignatureAlgorithm = x509.SHA256WithRSA
+		case certutil.ECPrivateKey:
+			certTemplate.SignatureAlgorithm = x509.ECDSAWithSHA256
+		}
+
+		caCert := creationInfo.SigningBundle.Certificate
+
+		certBytes, err = x509.CreateCertificate(rand.Reader, certTemplate, caCert, clientPrivKey.Public(), creationInfo.SigningBundle.PrivateKey)
+	} else {
+		// Creating a self-signed root
+		if creationInfo.MaxPathLength == 0 {
+			certTemplate.MaxPathLen = 0
+			certTemplate.MaxPathLenZero = true
+		} else {
+			certTemplate.MaxPathLen = creationInfo.MaxPathLength
+		}
+
+		switch creationInfo.KeyType {
+		case "rsa":
+			certTemplate.SignatureAlgorithm = x509.SHA256WithRSA
+		case "ec":
+			certTemplate.SignatureAlgorithm = x509.ECDSAWithSHA256
+		}
+
+		certTemplate.IsCA = true
+		certTemplate.KeyUsage = x509.KeyUsage(certTemplate.KeyUsage | x509.KeyUsageCertSign | x509.KeyUsageCRLSign)
+		certTemplate.ExtKeyUsage = append(certTemplate.ExtKeyUsage, x509.ExtKeyUsageOCSPSigning)
+		certBytes, err = x509.CreateCertificate(rand.Reader, certTemplate, certTemplate, clientPrivKey.Public(), clientPrivKey)
+	}
+
+	if err != nil {
+		return nil, certutil.InternalError{Err: fmt.Sprintf("unable to create certificate: %s", err)}
+	}
+
+	result.CertificateBytes = certBytes
+	result.Certificate, err = x509.ParseCertificate(certBytes)
+	if err != nil {
+		return nil, certutil.InternalError{Err: fmt.Sprintf("unable to parse created certificate: %s", err)}
+	}
+
+	if creationInfo.SigningBundle != nil {
+		result.IssuingCABytes = creationInfo.SigningBundle.CertificateBytes
+		result.IssuingCA = creationInfo.SigningBundle.Certificate
+	} else {
+		result.IssuingCABytes = result.CertificateBytes
+		result.IssuingCA = result.Certificate
+	}
+
+	return result, nil
+}
+
+// Creates a CSR. This is currently only meant for use when
+// generating an intermediate certificate.
+func createCSR(creationInfo *certCreationBundle) (*certutil.ParsedCSRBundle, error) {
```

Comment:
Lets rename to clarify. Also lets try to stay to the `godoc` style, useful when using the doc generators.